container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/persistent_directories.go

```diff
@@ -0,0 +1,75 @@
+package user_service_functions
+
+import (
+	"context"
+	"github.com/kurtosis-tech/kurtosis/container-engine-lib/lib/backend_impls/docker/docker_manager"
+	"github.com/kurtosis-tech/kurtosis/container-engine-lib/lib/backend_impls/docker/object_attributes_provider"
+	"github.com/kurtosis-tech/kurtosis/container-engine-lib/lib/backend_interface/objects/service"
+	"github.com/kurtosis-tech/kurtosis/container-engine-lib/lib/backend_interface/objects/service_directory"
+	"github.com/kurtosis-tech/stacktrace"
+	"github.com/sirupsen/logrus"
+)
+
+func getOrCreatePersistentDirectories(
+	ctx context.Context,
+	serviceUuid service.ServiceUUID,
+	objAttrsProvider object_attributes_provider.DockerEnclaveObjectAttributesProvider,
+	serviceMountpointsToPersistentKey map[string]service_directory.DirectoryPersistentKey,
+	dockerManager *docker_manager.DockerManager,
+) (map[string]string, error) {
+	shouldDeleteVolumes := true
+	result := map[string]string{}
+
+	for serviceDirpath, persistentKey := range serviceMountpointsToPersistentKey {
+		volumeAttrs, err := objAttrsProvider.ForSinglePersistentDirectoryVolume(serviceUuid, persistentKey)
+		if err != nil {
+			return nil, stacktrace.Propagate(err, "Error creating persistent directory labels for '%s'", persistentKey)
+		}
+
+		volumeName := volumeAttrs.GetName().GetString()
+		volumeLabelsStrs := map[string]string{}
+		for key, value := range volumeAttrs.GetLabels() {
+			volumeLabelsStrs[key.GetString()] = value.GetString()
+		}
+
+		potentiallyExistingVolumes, err := dockerManager.GetVolumesByName(ctx, volumeName)
+		if err != nil {
+			return nil, stacktrace.Propagate(err, "An error occurred checking for persistent volume existence")
+		}
+		if len(potentiallyExistingVolumes) == 1 {
+			result[serviceDirpath] = volumeName
+			continue
+		} else if len(potentiallyExistingVolumes) > 1 {
+			return nil, stacktrace.NewError("More than one volume with name '%s' exists in docker. This is unexpected", volumeName)
+		}
+
+		if err = dockerManager.CreateVolume(ctx, volumeName, volumeLabelsStrs); err != nil {
+			return nil, stacktrace.Propagate(
+				err,
+				"An error occurred creating persistent directory volume '%s' for service '%v'",
+				persistentKey,
+				serviceUuid,
+			)
+		}
+		result[volumeName] = serviceDirpath
+	}
+
+	defer func() {
+		if !shouldDeleteVolumes {
+			return
+		}
+		for _, volumeNameStr := range result {
```

Comment:
`result` could contain previous existing volumes right? Is it ok to remove those also?

Reply:
Yes you're right, I guess we should probable leave the already existing volume untouched in this case. I'll change that. Good catch!